### ISSUE TITLE
Application crashes when setting up swapchain on Vulkan SDK version of MoltenVK

### ISSUE THREAD

**Issue by marchingcubes** [KhronosGroup/MoltenVK#141]
When attempting to set up a swapchain on an OS X 10.11.6 system, I get the crash below.

I note that this same code works fine with MoltenVK 0.18, the version I was using for testing.

My swapchain is requesting VK_PRESENT_MODE_IMMEDIATE_KHR.

```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[CAMetalLayer setDisplaySyncEnabled:]: unrecognized selector sent to instance 0x7fa44051dbd0'
terminating with uncaught exception of type NSException
abort() called

Application Specific Backtrace 1:
0   CoreFoundation                      0x00007fff9107e452 __exceptionPreprocess + 178
1   libobjc.A.dylib                     0x00007fff9201af7e objc_exception_throw + 48
2   CoreFoundation                      0x00007fff910e818d -[NSObject(NSObject) doesNotRecognizeSelector:] + 205
3   CoreFoundation                      0x00007fff90fee4c1 ___forwarding___ + 1009
4   CoreFoundation                      0x00007fff90fee048 _CF_forwarding_prep_0 + 120
5   libMoltenVK.dylib                   0x000000010bc7e5aa _ZN12MVKSwapchainC2EP9MVKDevicePK24VkSwapchainCreateInfoKHR + 202
6   libMoltenVK.dylib                   0x000000010bc6926b _ZN9MVKDevice15createSwapchainEPK24VkSwapchainCreateInfoKHRPK21VkAllocationCallbacks + 43
7   libMoltenVK.dylib                   0x000000010bc4bdd2 vkCreateSwapchainKHR + 18
8   VulkanEngine                        0x000000010baf1821 _ZN12VulkanEngine15createSwapchainEv + 241
9   VulkanEngine                        0x000000010baf083e _ZN12VulkanEngine29createStaticRendererResourcesEv + 78
10  VulkanEngine                        0x000000010bae9db1 main + 2465
11  libdyld.dylib                       0x00007fff913e35ad start + 1

```

**Comment by marchingcubes** [comment]
I should also note that the application renders fine with VK_PRESENT_MODE_FIFO_KHR used instead, so possibly it is just not sending the 'false' bool parameter when it should?

**Comment by billhollings** [comment]
@marchingcubes 

`vkGetPhysicalDeviceSurfacePresentModesKHR()` will not include `VK_PRESENT_MODE_IMMEDIATE_KHR` unless it is supported on the device, and you should not create a swap chain with any present mode not included in that list.

I've fixed the crash in PR #142...and **MoltenVK** now avoids the crash if `VK_PRESENT_MODE_IMMEDIATE_KHR` is used incorrectly.

The crash did not appear on previous versions of **MoltenVK** because `VK_PRESENT_MODE_IMMEDIATE_KHR ` was not supported at all...and was simply ignored. Now it is supported if it is available (macOS 10.13 only).

**Comment by but0n** [comment]
@billhollings 
Hi, 
I have  same problems when setting up swapchain.
My code is [here](https://gist.github.com/but0n/e5f2f9f680f756d397f8b344703d818c).
I tried to modify the [presentMode](https://gist.github.com/but0n/e5f2f9f680f756d397f8b344703d818c#file-helloworld-c-L250) of swapchain create info, but it is still crashing no matter what presentMode I selected.

Here is the console messages:
```bash
SDL Extension Count : 2


	Extension 0: VK_KHR_surface


	Extension 1: VK_KHR_surface

[mvk-info] MoltenVK version 1.0.12. Vulkan version 1.0.77.
[mvk-info] GPU device:
		model: Intel(R) Iris(TM) Plus Graphics 640
		type: Integrated
		vendorID: 0x8086
		deviceID: 0x5926
		pipelineCacheUUID: 00000000-0000-0000-0000-27130000271C
	supports the following Metal Feature Sets:
		macOS GPU Family 1 v3
		macOS GPU Family 1 v2
		macOS GPU Family 1 v1

Physical devices Count : 1

	Driver Version: 10012
	Device Name:    Intel(R) Iris(TM) Plus Graphics 640
	Device Type:    1
	API Version:    1.0.77
GPU[0] Count of Queues: 8
Supported operationg on this queue:
		 Graphics @ 0
		 Compute @ 0
		 Transfer @ 0
[mvk-info] Created VkDevice to run on GPU Intel(R) Iris(TM) Plus Graphics 640

GPU supported? Yes

 - support extension amount:  5
	 Name: VK_MVK_moltenvk
	 Name: VK_KHR_swapchain
	 Name: VK_KHR_surface
	 Name: VK_AMD_negative_viewport_height
	 Name: VK_MVK_macos_surface
Resolution:
	Width - 640
	Height - 480
Max Resolution:
	Width - 640
	Height - 480
Physical Device Surface Formats count: 3

	 Format 0:
		 * format: 2c
		 * colorSpace: 0
	 Format 1:
		 * format: 32
		 * colorSpace: 0
	 Format 2:
		 * format: 61
		 * colorSpace: 0

Physical Device Surface Present Modes count: 2

	 Present Modes 0:
		 * mode: 2
	 Present Modes 1:
		 * mode: 0
2018-07-22 20:47:11.513 helloVulkan[15338:838908] -[NSTaggedPointerString setDevice:]: unrecognized selector sent to instance 0x5
2018-07-22 20:47:11.514 helloVulkan[15338:838908] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSTaggedPointerString setDevice:]: unrecognized selector sent to instance 0x5'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff4fbb9a17 __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff7b9bea18 objc_exception_throw + 48
	2   CoreFoundation                      0x00007fff4fc414fc -[NSObject(NSObject) doesNotRecognizeSelector:] + 132
	3   CoreFoundation                      0x00007fff4fb48992 ___forwarding___ + 1490
	4   CoreFoundation                      0x00007fff4fb48338 _CF_forwarding_prep_0 + 120
	5   libMoltenVK.dylib                   0x000000010c229500 _ZN12MVKSwapchainC2EP9MVKDevicePK24VkSwapchainCreateInfoKHR + 112
	6   libMoltenVK.dylib                   0x000000010c21327b _ZN9MVKDevice15createSwapchainEPK24VkSwapchainCreateInfoKHRPK21VkAllocationCallbacks + 43
	7   libMoltenVK.dylib                   0x000000010c1f6e82 vkCreateSwapchainKHR + 18
	8   libvulkan.1.dylib                   0x0000000106b15bc5 vkCreateSwapchainKHR + 69
	9   helloVulkan                         0x0000000106ae893c getDevices + 2220
	10  helloVulkan                         0x0000000106ae89b6 main + 54
	11  libdyld.dylib                       0x00007fff7c99a3ed start + 1
	12  ???                                 0x0000000000000001 0x0 + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
[1]    15338 abort      ./helloVulkan
```

Is there something wrong with my configuration?

Thanks.

**Comment by billhollings** [comment]
@but0n 

It looks like the surface you are submitting in `VkSwapchainCreateInfoKHR ::surface` is not valid...and specifically...does not contain a valid `CAMetalLayer`.

In your code, it looks like this surface is created by `SDL_Vulkan_CreateSurface()`...which is not a MoltenVK function...and I assume is part of SDL.

Are you able to determine what that `SDL_Vulkan_CreateSurface()` function is doing?

**Comment by mrtrizer** [comment]
Versions of components:
2fd5a24ec4a6df303b2155b3f85b6b8c1d56f6c0 : https://github.com/KhronosGroup/Vulkan-Headers.git
98d5fac6c461959e64bdbe27c710a93215147dd1 : https://github.com/KhronosGroup/Vulkan-Loader.git
bdd65da74e39e0953a222275de9033543807ccdc : https://github.com/KhronosGroup/Vulkan-ValidationLayers.git
3796b651c94a8bfb383bf42384d6dabc7e60908d : https://github.com/KhronosGroup/Vulkan-Tools.git
26c7b69a3fbee8125ff6072f766c10147d00754d : https://github.com/KhronosGroup/MoltenVK.git
91ac4290bcf2cb930b4fb0981f09c00c0b6797e1 : https://github.com/KhronosGroup/glslang.git
9bfe0eb25e3dfdf4f3fd86ab6c0cda009c9bd661 : https://github.com/KhronosGroup/SPIRV-Tools
d5b2e1255f706ce1f88812217e9a554f299848af : https://github.com/KhronosGroup/SPIRV-Headers
34014886e338e57940eef32859622d9ca31f7f53 : https://github.com/KhronosGroup/SPIRV-Cross.git
c15f914b0f62bf24c58fb201ad6f4b10ecaa122c : https://github.com/google/shaderc.git
f1aa87e59828fd2f71790b69d0c33f81ed4de2e1 : https://github.com/KhronosGroup/Vulkan-Docs.git

I'm struggling with similar issue in my experiments.
When I link MoltenVK via Vulkan loader as ICU driver, it falls with trace:
```
2018-11-25 13:42:43.261415+0300 TestApp[40092:3235772] MessageTracer: Falling back to default whitelist
2018-11-25 13:42:43.278224+0300 TestApp[40092:3235772] -[NSTaggedPointerString setDevice:]: unrecognized selector sent to instance 0x7fff3e110005
2018-11-25 13:42:43.281313+0300 TestApp[40092:3235772] [General] An uncaught exception was raised
2018-11-25 13:42:43.281350+0300 TestApp[40092:3235772] [General] -[NSTaggedPointerString setDevice:]: unrecognized selector sent to instance 0x7fff3e110005
2018-11-25 13:42:43.281443+0300 TestApp[40092:3235772] [General] (
    0   CoreFoundation                      0x00007fff3ff6f2db __exceptionPreprocess + 171
    1   libobjc.A.dylib                     0x00007fff6710cc76 objc_exception_throw + 48
    2   CoreFoundation                      0x00007fff40007db4 -[NSObject(NSObject) doesNotRecognizeSelector:] + 132
    3   CoreFoundation                      0x00007fff3fee5820 ___forwarding___ + 1456
    4   CoreFoundation                      0x00007fff3fee51e8 _CF_forwarding_prep_0 + 120
    5   libMoltenVK.dylib                   0x000000010284708f _ZN12MVKSwapchainC2EP9MVKDevicePK24VkSwapchainCreateInfoKHR + 111
    6   libMoltenVK.dylib                   0x00000001028304cb _ZN9MVKDevice15createSwapchainEPK24VkSwapchainCr
eateInfoKHRPK21VkAllocationCallbacks + 43
    7   libMoltenVK.dylib                   0x000000010280ff4b vkCreateSwapchainKHR + 27
    8   TestApp                             0x0000000100009686 _Z15createSwapchainP18VkPhysicalDevice_TP10VkDevice_TP14VkSurfaceKHR_TRK17SwapchainSettings + 358
    9   TestApp                             0x000000010000daf8 main + 4104
    10  libdyld.dylib                       0x00007fff67d26015 start + 1
)
2018-11-25 13:42:43.281921+0300 TestApp[40092:3235772] *** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '-[NSTaggedPointerString setDevice:]: unrecognized selector sent to instance 0x7fff3e110005'
*** First throw call stack:
(
    0   CoreFoundation                      0x00007fff3ff6f2db __exceptionPreprocess + 171
    1   libobjc.A.dylib                     0x00007fff6710cc76 objc_exception_throw + 48
    2   CoreFoundation                      0x00007fff40007db4 -[NSObject(NSObject) doesNotRecognizeSelector:] + 132
    3   CoreFoundation                      0x00007fff3fee5820 ___forwarding___ + 1456
    4   CoreFoundation                      0x00007fff3fee51e8 _CF_forwarding_prep_0 + 120
    5   libMoltenVK.dylib                   0x000000010284708f _ZN12MVKSwapchainC2EP9MVKDevicePK24VkSwapchainCreateInfoKHR + 111
    6   libMoltenVK.dylib                   0x00000001028304cb _ZN9MVKDevice15createSwapchainEPK24VkSwapchainCreateInfoKHRPK21VkAllocationCallbacks + 43
    7   libMoltenVK
.dylib                   0x000000010280ff4b vkCreateSwapchainKHR + 27
    8   TestApp                             0x0000000100009686 _Z15createSwapchainP18VkPhysicalDevice_TP10VkDevice_TP14VkSurfaceKHR_TRK17SwapchainSettings + 358
    9   TestApp                             0x000000010000daf8 main + 4104
    10  libdyld.dylib                       0x00007fff67d26015 start + 1
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```
Avaliable layers and extensions:
```
Validation layers:
    VK_LAYER_GOOGLE_unique_objects
    VK_LAYER_GOOGLE_threading
    VK_LAYER_LUNARG_standard_validation
    VK_LAYER_LUNARG_core_validation
    VK_LAYER_LUNARG_parameter_validation
    VK_LAYER_LUNARG_object_tracker
Required extensions for instance:
    VK_KHR_surface
    VK_MVK_macos_surface
available extensions:
    VK_KHR_get_physical_device_properties2
    VK_KHR_surface
    VK_MVK_macos_surface
    VK_EXT_debug_report
    VK_EXT_debug_utils
Graphics family index: 0
Present family index: 0
```
Main source with vulkan initialization:
https://github.com/mrtrizer/VulkanTriangleMacOS/blob/master/src/main.cpp
Surface init ObjC:
https://github.com/mrtrizer/VulkanTriangleMacOS/blob/master/src/macOSInterface.m
CMakeLists:
https://github.com/mrtrizer/VulkanTriangleMacOS/blob/master/generated/cmake/CMakeLists.txt

**Comment by mrtrizer** [comment]
Okay, I've finally succeed. I've rolled back to 1.1.77.0 and after that I started seeing messages from debug layout. It appeared that I didn't enable `VK_KHR_swapchain` extension in `VkDeviceCreateInfo`. After enabling I've quickly gone forward.
Moreover, I've used your cube app example to initialize window properly.

**Comment by mrtrizer** [comment]
I've finally succeed. I've rolled back to 1.1.77.0 and debug layouts started working. It's appeared that I didn't enable `VK_KHR_swapchain` in `VkDeviceCreateInfo`. After enabling it I quickly gone forward.
I used your cube example to initalize NSWindow with layout backed NSView right way and finally seen the triangle. I didn't think this gonna take so much time :)
(if anybody intereseted in pure initialization from code, refer this example: https://github.com/mrtrizer/VulkanTriangleMacOS/blob/master/src/macOSInterface.m)
So, the main problem I see now is that Vulkan didn't show me output from debug layers. But seems the problem is not on side of MoltenVK. Maybe tutorial I followed is outdate or something like this.
Anyway, thanks for Molten! 👍
